### PR TITLE
Parallelize Soroban pre-apply step.

### PR DESCRIPTION
**Pull request by dmkozh**
# Description

While the current logic intertwines reads and writes, in fact it can be cleanly separated into a read-only validation step, and a sequential commit step that simply bumps the sequence numbers and removes pre-authorized tx signers. This is possible because that while the writes change the entries that take part in validation, none of these changes are relevant during the validation. Specifically, sequence number bump is only observable by a single transaction (the one that has the respective account as a source), and the pre-authorized tx signer by definition belongs to a single transaction. There is also a subtle caveat to the latter operation: it increases the available balance of the signer owner (or its sponsor), but since at the pre-apply time the fees have already been charged, we're only checking that the account available balance is non-negative, which is an invariant that must always hold in the current protocol.

The change is not protocol-gated because it's not a protocol change for the *current* protocol. It was technically a protocol change prior to p26 where we had a bug that actually did allow overcharging the fee bump source accounts and thus making their available balance to go negative. However, the bug has been fixed without the behavior ever triggering on-chain, and thus this replay-only behavior change should be non-observable.

This change significantly speeds up the pre-apply step. On the local high TPL benchmarks I'm getting 30-60ms improvement locally compared to the main branch version.

This PR also adds test coverage for the edge case pre-apply interactions that weren't covered well (or at all), such as pre-auth signer removal, tx invalidation due to sequence bumps etc. This also contains a randomized acceptance test that ensures that parallel pre-apply logic is deterministic across any number of worker threads.

# Checklist
- [ ] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [ ] Rebased on top of master (no merge commits)
- [ ] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [ ] Compiles
- [ ] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/TransactionFrame.cpp [hunk 1]
```diff
@@ -2116,67 +2141,92 @@ TransactionFrame::commonPreApply(bool chargeFee, AppConnector& app,
...
+    AppConnector& app, AbstractLedgerTxn& ltx, TransactionMetaBuilder& meta,
+    MutableTransactionResultBase const& txResult) const
+{
+    ZoneScoped;
+    releaseAssert(threadIsMain() ||
+                  app.threadIsType(Application::ThreadType::APPLY));
+    try
+    {
+        LedgerTxn ltxTx(ltx);
+        if (shouldUpdateSeqNumInPreApply(txResult))
```
Comment (sisuresh) [inline src/transactions/TransactionFrame.cpp:2171]: Should we assert that the codes aren't ones that we wouldn't expect like `txINSUFFICIENT_FEE`?

Comment (dmkozh) [reply src/transactions/TransactionFrame.cpp:2171]: Not sure I understand, I'm not sure if we can make a safe assumption on which codes are truly unexpected, and I don't believe we had such an assertion before.

Comment (sisuresh) [reply src/transactions/TransactionFrame.cpp:2171]: In `commonValid`, we only set `res = ValidationType::kInvalidUpdateSeqNum` if `commonValidPreSeqNum` does not set an error, so in the old code we would not consume the seq num in those error scenarios right? Here we would, which would be a difference in behavior IIUC. We shouldn't be able to get to this line due to tx set validation, but there could be a bug. You're right that the previous path didn't assert, which I think would be too strict.

Comment (dmkozh) [reply src/transactions/TransactionFrame.cpp:2171]: > so in the old code we would not consume the seq num in those error scenarios right? 

That's what we handle `shouldUpdateSeqNumInPreApply`, unless I'm mis-understanding. We only bump sequence number when the error is not bad_seq/no_account, which corresponds to `ValidationType::kInvalidUpdateSeqNum` in the old flow.

Comment (sisuresh) [reply src/transactions/TransactionFrame.cpp:2171]: We concluded that this does not need to be addressed.

File: src/util/BatchExecutor.cpp [hunk 2]
```diff
@@ -210,6 +210,60 @@ BatchExecutor::pinWorker(size_t index)
...
+void
+BatchExecutor::executeBatchOverRanges(
+    size_t count, size_t numTasks,
+    std::function<void(size_t, size_t, size_t)> const& work)
+{
+    if (numTasks <= 1 || count < numTasks)
+    {
+        work(0, count, 0);
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/util/BatchExecutor.cpp:222]: That's actually by design, updated the comments.

File: src/transactions/ParallelApplyUtils.cpp [hunk 3]
```diff
@@ -353,34 +369,77 @@ GlobalParallelApplyLedgerState::
...
+
+    auto& batchExecutor = app.getBatchExecutor();
+    batchExecutor.executeBatchOverRanges(
+        txBundles.size(), batchExecutor.preferredTaskCount(),
+        [&](size_t begin, size_t end, size_t) {
+            // NB: mLCLApplyView is not thread-safe, so we need to copy it into
+            // a thread-local view.
+            CheckValidLedgerViewWrapper ledgerView(
+                std::make_unique<SorobanPreApplyLedgerView>(header, ltx,
+                                                            mLCLApplyView));
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply src/transactions/ParallelApplyUtils.cpp:434]: getNewestVersionBelowRoot is thread-safe in read-only context.

**Review by SirTyson — DISMISSED** [review]
Overall LGTM. I am most familiar with the snapshot interface, it might be nice for someone with more experience in the tx subsystem to take a look as well, but I can't see any form of divergence for current protocol.

I still have a bit of a reservation regarding multi-threaded structs accessing the ltx object. I agree it's all correct, but I almost want some abstraction like `ThreadSafeLtx `with a limited set of const, thread safe map accessors, and some way of "deactivating" the parent ltx object such that we can ensure there are no races. That being said, we already have many, many layers of abstraction in the snapshot and state access interface, and we only use `getNewestVersionBelowRoot`, so maybe it's fine as is.

File: src/transactions/ParallelApplyUtils.cpp [hunk 4]
```diff
@@ -353,34 +369,98 @@ GlobalParallelApplyLedgerState::
...
+    //   because any given transaction in a ledger is unique, and decreasing the
+    //   sub-entry count of a source/sponsor account is not relevant at that
+    //   point, as the fees have already been successfully charged.
+
+    auto header =
+        std::make_shared<LedgerHeader const>(ltx.loadHeader().current());
+    readOnlyParallelPreApply(app, txBundles, header, ltx);
+    commitPreParallelApplyWrites(app, ltx, txBundles);
+
+    for (auto const& txBundle : txBundles)
```
Comment (SirTyson) [inline src/transactions/ParallelApplyUtils.cpp]: Nit: `auto const*`

File: src/ledger/ImmutableLedgerView.h [hunk 5]
```diff
@@ -219,6 +220,45 @@ class ApplyLedgerView : private ImmutableLedgerView,
...
+// It's a thin wrapper around the entries updated so far in the current ledger,
+// and the LCL view, which allows the pre-apply phase to observe the changes
+// that happened in the classic phase.
+//
+// Lookups are first attempted among the updated entries, and only then in the
+// LCL view.
+class SorobanPreApplyLedgerView : public AbstractLedgerView
+{
+  public:
+    using UpdatedEntryGetter =
```
Comment (SirTyson) [inline src/ledger/ImmutableLedgerView.h:240]: Can we add a comment explaining this interface? It's a fairly verbose and indirect type.

**Review by SirTyson — DISMISSED** [review]
Thanks for the hadening around the LTX interface! Two small nits.

File: src/transactions/ParallelApplyUtils.cpp (hunk 4 above)
Comment (dmkozh) [reply src/transactions/ParallelApplyUtils.cpp]: Done.,

File: src/ledger/ImmutableLedgerView.h (hunk 5 above)
Comment (dmkozh) [reply src/ledger/ImmutableLedgerView.h:240]: Done.

**Review by sisuresh — APPROVED** [review]
